src/write.rs

```diff
@@ -476,7 +483,177 @@ where
+                // while attempting to combine them with the next op,
+                // cut short when an attempt fails due to encountering a dependency (e.g. clear then set).
+
+                // used to avoid walking more of the oplog than necessary.
+                let mut rev_dirty_range = 0usize..0;
+                for mut next in ops {
+                    // used to avoid linear insertion time in case of predominantly independent ops.
+                    let mut range_remaining = T::MAX_COMPRESS_RANGE;
+                    // used to more efficiently insert next if possible
+                    let mut none: Option<(usize, &mut Option<O>)> = None;
+                    // rev-iterate all unpublished and potentially non-none ops already in the oplog
+                    for (prev_rev_idx, prev_loc) in {
+                        #[cfg(test)]
+                        {
+                            // While testing, make very very sure we don't skip any Some.
+                            let none_rev_idx = rev_dirty_range.start.saturating_sub(1);
+                            self.oplog
+                                .iter_mut()
+                                .skip(self.swap_index) // only consider the fresh part of the oplog
+                                .rev() // We need to walk it in reverse
+                                .enumerate() // we need the reverse index for the rev_dirty_range optimization
+                                .skip_while(move |(rev_idx, loc)| {
+                                    if *rev_idx < none_rev_idx {
+                                        debug_assert!(loc.is_none(), "We never skip over Some.",);
+                                        true
+                                    } else {
+                                        debug_assert!(
+                                            *rev_idx == none_rev_idx || loc.is_some(),
+                                            "We either stop on some or the last none",
+                                        );
+                                        false
+                                    }
+                                }) // skip nones at the back (except one for efficient insertion)
+                        }
+                        #[cfg(not(test))]
+                        {
+                            self.oplog
+                                .iter_mut()
+                                .skip(self.swap_index) // only consider the fresh part of the oplog
+                                .rev() // We need to walk it in reverse
+                                .enumerate() // we need the reverse index for the rev_dirty_range optimization
+                                .skip(rev_dirty_range.start.saturating_sub(1)) // skip nones at the back (except one for efficient insertion)
+                        }
+                    } {
+                        if let Some(prev) = prev_loc.as_mut() {
```

Comment:
Seems weird for it to be possible for this to be `None`. Can't that only happen if the oplog (after `self.swap_index`) is empty?

Reply:
It is None if a previous iteration has left behind a gap.

[Some, None, Some, None, None] Would completely skip last None, start on and remember second-to-last None, try to compress with last Some, encounter and remember first None, try to compress with first Some